Better RT refresh & mantainence

aarshkshah1992 opened the pull request:
@Stebalien 
For #283, #434  & https://github.com/libp2p/go-libp2p/issues/779.

Corresponding Routing Table PR:
https://github.com/libp2p/go-libp2p-kbucket/pull/54

aarshkshah1992 commented:
ping @Stebalien

Stebalien commented on dht.go:
```diff
@@ -179,9 +189,26 @@ func NewDHTClient(ctx context.Context, h host.Host, dstore ds.Batching) *IpfsDHT
+			pvLogger.Errorf("failed to connect to peer %s for validation, err=%s", p, err)
+			return false
+		}
+		fmt.Print("Returning true")
```
debug

aarshkshah1992 replied:
This is done.

Stebalien commented on dht.go:
```diff
@@ -179,9 +189,26 @@ func NewDHTClient(ctx context.Context, h host.Host, dstore ds.Batching) *IpfsDHT
 	self := kb.ConvertPeerID(h.ID())
-	rt := kb.NewRoutingTable(bucketSize, self, time.Minute, h.Peerstore())
+	// construct the routing table with a peer validation function
+	pvLogger := logging.Logger("RT peer validation")
```
should be global

(and should have a name like "dht/rt")

aarshkshah1992 replied:
This is done.

Stebalien commented on dht_bootstrap.go:
```diff
@@ -88,14 +144,52 @@ func (dht *IpfsDHT) startRefreshing() error {
+				switch evt.(type) {
+				case event.EvtLocalAddressesUpdated:
+					// our address has changed, trigger a self walk so our closet peers know about it
+					res := make(chan error, 1)
```
I believe we can send a nil channel if we don't want a response (but double check that).

aarshkshah1992 replied:
We can indeed. This is done.

Stebalien commented on dht_bootstrap.go:
```diff
@@ -35,6 +37,60 @@ func init() {
+// startSelfLookup starts a go-routine that listens for requests to trigger a self walk on a dedicated channel
+// and then sends the error status back on the error channel sent along with the request.
+// if multiple callers "simultaneously" ask for a self walk, it performs ONLY one self walk and sends the same error status to all of them.
+func (dht *IpfsDHT) startSelfLookup() error {
```
Revisiting this, I'm not sure if we need something quite this complicated. We can probably get away with just calling `selfWalk` from within both the refresh function and the "address changed" loop.

(although we'll need to do it asynchronously inside the "address changed" loop)

Note: this code looks fine. I'm just not sure if the complexity is warranted.

willscott commented on dht_bootstrap.go:
```diff
@@ -17,7 +19,7 @@ var DefaultBootstrapPeers []multiaddr.Multiaddr
 // Minimum number of peers in the routing table. If we drop below this and we
 // see a new peer, we trigger a bootstrap round.
-var minRTRefreshThreshold = 4
+var minRTRefreshThreshold = 20
```
this seems like a high minimum for triggering new bootstraps. was this for testing, or intentional?

aarshkshah1992 replied:
This is intentional. The current interval for refresh is too much (1 hour) & we want to be more proactive in doing refresh when needed rather than doing them to often without needing to do them. But really, we can always tune these params if we run into problems later.

willscott commented on dht.go:
```diff
@@ -710,9 +750,9 @@ func (dht *IpfsDHT) handleProtocolChanges(proc goprocess.Process) {
 				// TODO: discuss how to handle this case
```
should this also trigger an update?

aarshkshah1992 replied:
@willscott Great spot ! It will after [this code](https://github.com/libp2p/go-libp2p-kad-dht/blob/c649402191b967192873b158f5950734110c43cc/subscriber_notifee.go#L162) in Adin's PR goes in.

willscott commented on dht.go:
```diff
@@ -193,31 +219,45 @@ func makeDHT(ctx context.Context, h host.Host, dstore ds.Batching, protocols []p
 
+	dht.subscriptions.everPeerAddressChanged, err = h.EventBus().Subscribe(&event.EvtLocalAddressesUpdated{}, eventbus.BufSize(256))
+	if err != nil {
+		logger.Errorf("dht not sybscribed to peer address changed event; err=%s ", err)
```
```suggestion
		logger.Errorf("dht not subscribed to peer address changed event; err=%s ", err)
```

aarshkshah1992 replied:
@willscott Oh that's just how we spell it down here. It's a very us thing.

willscott commented on dht_bootstrap.go:
```diff
@@ -35,6 +37,60 @@ func init() {
+			}
+
+			// Batch multiple refresh requests if they're all waiting at the same time.
+		collectWaiting:
```
can this loop of collecting all waiting channels be broken out into a separate function?

aarshkshah1992 replied:
This is done.

willscott commented on dht_bootstrap.go:
```diff
@@ -88,14 +144,51 @@ func (dht *IpfsDHT) startRefreshing() error {
+					return
+				}
+				switch evt.(type) {
+				case event.EvtLocalAddressesUpdated:
```
since we only care about one type here, can we do this as
```if _, ok := evt.(event.EvtLocalAddressesUpdated); ok { ... }```

aarshkshah1992 replied:
This is Done.

aarshkshah1992 commented:
Status Update @Stebalien @aschmahmann @willscott 

I have addressed review comments that I've gotten till now. Will rebase this on the cypress branch once Adin's work on switching between client/server goes in and ask you guys to take one final look.

aarshkshah1992 commented:
Closing in favour of #482
